src/Database/LSMTree/Internal/Run/Index/Compact/Construction.hs

```diff
@@ -287,17 +278,15 @@ unsafeEnd mci@MCompactIndex{..} = do
     -- combinations, so we have to fill in the remainder of the rangerfinder
     -- vector.
     fillRangeFinderToEnd mci
-    fcRangeFinder <- VU.unsafeFreeze mciRangeFinder
-
-    let completePrimary = PrimaryArray $ VU.concat . reverse $ chunksPrimary
-    let fcClashes = VU.concat . reverse $ chunksClashes
-    let fcLargerThanPage = VU.concat . reverse $ chunksLargerThanPage
-    fcTieBreaker <- readSTRef mciTieBreaker
+    ciRangeFinder <- VU.unsafeFreeze mciRangeFinder
```

Comment:
If we were to use safe `freeze` here then this whole `unsafeEnd` could in fact be safe.

All the other uses of `unsafeFreeze` are actually safe since the result gets copied (`VU.concat`) before being returned.

Note `VU.concat` probably has a non-copying optimisation for a single element input, so one would need a little utility function that copies for a single chunk and uses concat for all other cases.

Reply:
I'm leaning towards keeping it unsafe. The usage pattern of these mutable builders seems pretty clear and I don't see a reason to keep using them after calling `unsafeEnd`/`unsafeFinalise`.
On the contrary, we might want to switch to `Bloom.unsafeFreeze` in `RunAcc` to avoid copying the bloom filter.

Reply:
I switched to `Bloom.unsafeFreeze`, but let me know if you disagree.

Reply:
Users won't have access to `unsafeEnd` and co, but of course we as developers can still make a mistake :P In any case, I think it's fine, though I don't have a strong preference